File: android/app/src/main/kotlin/run/nuri/getagrip/ui/share/RoutineImportSheet.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.share

import run.nuri.getagrip.ui.units.WeightUnits

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch
import run.nuri.getagrip.engine.FingerSet
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.engine.PlanMath
import run.nuri.getagrip.engine.RoutineDraft
import run.nuri.getagrip.engine.RoutineSummary
import run.nuri.getagrip.engine.SessionPlan
import run.nuri.getagrip.engine.SetPlan
import run.nuri.getagrip.store.LocalTemplateStore
import run.nuri.getagrip.ui.components.CapsLabel
import run.nuri.getagrip.ui.components.EdgeMark
import run.nuri.getagrip.ui.components.FingerGlyph
import run.nuri.getagrip.ui.components.PrimaryButton
import run.nuri.getagrip.ui.components.SecondaryButton
import run.nuri.getagrip.ui.components.tint
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics
import run.nuri.getagrip.ui.today.PlanRowFit

/// The other side of a QR code: somebody else's routine, read out in full, before it is yours.
///
/// A PREVIEW, not an editor: a routine you have not accepted is not yours to correct. Once it
/// lands, its card opens the builder on its own plan row.
///
/// **Percentage targets are NOT translated.** They resolve against the READER's maxes, per
/// hand, at session start — the reason the app prescribes fractions, so one code suits two
/// very different pairs of fingers. The footnotes cover where that is not the whole story.
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RoutineImportSheet(incoming: RoutineDraft, onClose: () -> Unit) {
    val palette = LocalGripPalette.current
    val templates = LocalTemplateStore.current
    val scope = rememberCoroutineScope()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)

    // Normalized HERE so the preview shows what will land (default name, emptied sets dropped,
    // inheritance consolidated). `normalized` is idempotent, so the store's pass costs nothing.
    val draft = remember(incoming) { incoming.normalized }
    // Folded ONCE: `RoutineSummary.previewing` walks the reps three times and mints an id, so a
    // computed read would re-fold a fifty-set routine per scroll frame and change identity.
    val summary = remember(draft) { RoutineSummary.previewing(draft) }
    val plan = draft.plan
    /// `executable`, like every fold: a set with no pulls is an emptied row, not a rest.
    val sets = remember(plan) { plan.executable.sets }

    /// The name this routine will actually LAND under (the store's `uniqueName`): two people keeping
    /// the default name is common, and "Daily no-hangs" becoming "Daily no-hangs 2" would be the
    /// preview and card disagreeing.
    ///
    /// TRANSLATION NOTE: iOS reads it synchronously; `plannedImportName` suspends here (Room), so
    /// it resolves into state, starting from the routine's own name so the header never flickers.
    var landingName by remember(draft) { mutableStateOf(plan.name) }
    LaunchedEffect(draft) { landingName = templates.plannedImportName(plan.name) }

    /// Whether THIS sheet's add failed — local, not the store's error field, or the sheet would
    /// open accused by an earlier unrelated failure.
    var saveFailed by remember { mutableStateOf(false) }
    /// An add in flight: the draft has no id, so a second tap inside the write added it twice.
    var adding by remember { mutableStateOf(false) }

    ModalBottomSheet(
        onDismissRequest = onClose,
        sheetState = sheetState,
        containerColor = palette.field,
        shape = RoundedCornerShape(topStart = Metrics.radiusSheet, topEnd = Metrics.radiusSheet),
    ) {
        Column(
            Modifier
                .verticalScroll(rememberScrollState())
                .padding(horizontal = Metrics.hPadding)
                .padding(bottom = Metrics.spacing),
            verticalArrangement = Arrangement.spacedBy(18.dp),
        ) {
            Text(
                tr("Shared routine"),
                style = MaterialTheme.typography.titleLarge,
                color = palette.inkPrimary,
            )

            Header(landingName, summary)
            PlanCard(sets, plan)
            RhythmCard(plan, draft, summary.setCount)
            Notes(sets, plan)

            if (saveFailed) {
                // STAYS OPEN on a rollback: dismissing loses the code too, and rescanning is somebody
                // else's phone away.
                Text(
                    tr("That routine couldn't be saved just now — nothing was added. Try again."),
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.Medium,
                    color = palette.alarm,
                )
            }

            PrimaryButton(tr("Add to my routines"), icon = Icons.Outlined.Add, enabled = !adding) {
                if (adding) return@PrimaryButton
                adding = true
                scope.launch {
                    try {
                        // The card appears on Today by itself: the routine list is store state.
                        if (templates.importRoutine(draft) != null) {
                            saveFailed = false
                            onClose()
                        } else {
                            // INLINE, and the store's copy is consumed: the global "Couldn't save" watches the same
                            // field, and one rollback stated twice reads as two.
                            saveFailed = true
                            templates.saveError = null
                        }
                    } finally {
                        adding = false
                    }
                }
            }

            // Quiet, never destructive-looking: declining costs nothing.
            SecondaryButton(tr("Not now"), modifier = Modifier.fillMaxWidth(), onClick = onClose)

            Spacer(Modifier.padding(bottom = 4.dp))
        }
    }
}

// MARK: - Identity

@Composable
private fun Header(landingName: String, summary: RoutineSummary) {
    val palette = LocalGripPalette.current
    Row(
        Modifier.clearAndSetSemantics {
            // The rung's colour is invisible to TalkBack and greyscale, so speak the number (as `RoutineCard`).
            contentDescription = buildString {
                append(landingName)
                append(". ")
                append(summary.metaLine)
                val peak = summary.peakIntensity
                if (peak != null) {
                    append(L10n.tr(". Peak target %d percent of max.", Math.round(peak * 100)))
                }
            }
        },
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.Top,
    ) {
        EdgeMark(
            fingers = summary.signatureFingers ?: FingerSet.four,
            rungTint = PlanMath.IntensityBand.band(summary.peakIntensity).tint(palette),
            modifier = Modifier.padding(top = 4.dp),
        )
        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(
                landingName,
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.SemiBold,
                color = palette.inkPrimary,
            )
            Text(
                summary.metaLine,
                style = MaterialTheme.typography.bodySmall.copy(fontFeatureSettings = "tnum"),
                color = palette.inkSecondary,
            )
        }
    }
}

// MARK: - The plan

@Composable
private fun PlanCard(sets: List<SetPlan>, plan: SessionPlan) {
    val palette = LocalGripPalette.current
    Surface(
        shape = RoundedCornerShape(Metrics.radiusCard),
        color = palette.card,
        modifier = Modifier.fillMaxWidth(),
    ) {
        Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(10.dp)) {
            CapsLabel(tr("WHAT YOU'LL PULL"))
            sets.forEachIndexed { index, set ->
                SetRow(set, plan)
                if (index < sets.size - 1) {
                    HorizontalDivider(color = palette.inkTertiary.copy(alpha = 0.14f))
                }
            }
        }
    }
}

/// Two layouts, forked at `.accessibility1` like `RoutineCard`'s plan row: side by side at the
/// type ceiling, glyph and count left the grip sentence a couple of dozen points. Big text
/// gets full-width words; the glyph is spare decoration.
@Composable
private fun SetRow(set: SetPlan, plan: SessionPlan) {
    val palette = LocalGripPalette.current
    val fontScale = LocalDensity.current.fontScale
    val reps = repText(set, plan)
    val detail = setDetailLine(set, plan)

    val spoken = buildString {
        append(set.grip.spoken)
        append(". ")
        append(reps)
        append(".")
        if (detail != null) append(" $detail.")
    }

    if (fontScale >= PlanRowFit.SENTENCE_FONT_SCALE) {
        Column(
            Modifier
                .fillMaxWidth()
                .clearAndSetSemantics { contentDescription = spoken },
            verticalArrangement = Arrangement.spacedBy(3.dp),
        ) {
            Text(
                set.grip.line,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
                color = palette.inkPrimary,
            )
            Text(
                reps,
                style = MaterialTheme.typography.bodySmall.copy(fontFeatureSettings = "tnum"),
                fontWeight = FontWeight.Medium,
                color = palette.inkSecondary,
            )
            if (detail != null) Detail(detail)
        }
        return
    }

    Row(
        Modifier
            .fillMaxWidth()
            .clearAndSetSemantics { contentDescription = spoken },
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.Top,
    ) {
        FingerGlyph(
            fingers = set.grip.fingers,
            position = set.grip.position,
            dot = 7.dp,
            gap = 3.dp,
            modifier = Modifier.padding(top = 3.dp),
        )
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            // No line limit: this sentence is what the screen exists to state.
            Text(
                set.grip.line,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
                color = palette.inkPrimary,
            )
            if (detail != null) Detail(detail)
        }
        Text(
            reps,
            style = MaterialTheme.typography.bodySmall.copy(fontFeatureSettings = "tnum"),
            fontWeight = FontWeight.Medium,
            color = palette.inkSecondary,
        )
    }
}

@Composable
private fun Detail(line: String) {
    Text(
        line,
        style = MaterialTheme.typography.bodySmall.copy(fontFeatureSettings = "tnum"),
        color = LocalGripPalette.current.inkSecondary,
    )
}

/// What this set does DIFFERENTLY — its own timing or load. A typed 40 kg band first met in
/// the runner is the surprise this sheet prevents, and a per-set hold explains why the
/// estimate disagrees with the rhythm line. Nothing for a set that inherits everything.
private fun setDetailLine(set: SetPlan, plan: SessionPlan): String? {
    val parts = mutableListOf<String>()
    set.holdSeconds?.let { parts.add(L10n.tr("%s hold", PlanMath.durationText(it))) }
    set.restSeconds?.let { parts.add(L10n.tr("%s rest", PlanMath.durationText(it))) }
    val kg = set.targetBand
    val percent = set.targetPercentBand
    if (kg != null) {
        parts.add(WeightUnits.tr("%s–%s kg target", kgText(kg.start), kgText(kg.endInclusive)))
    } else if (percent != null) {
        val lo = Math.round(percent.start * 100)
        val hi = Math.round(percent.endInclusive * 100)
        parts.add(
            if (lo == hi) L10n.tr("%d %% of max", hi) else L10n.tr("%d–%d %% of max", lo, hi)
        )
    }
    return parts.takeIf { it.isNotEmpty() }?.joinToString(" · ")
}

/// Display text, so the DEFAULT locale (as `PlanMath.bandText`); wire text is locale-free.
private fun kgText(kg: Double): String = WeightUnits.number(kg)

/// "6 per side" when hands alternate, "6 pulls" together. The count comes from
/// `PlanMath.repCount`, the one home of the `× sideCount` a silent factor of two would hide in.
private fun repText(set: SetPlan, plan: SessionPlan): String {
    if (plan.handMode.sideCount != 1) return L10n.tr("%d per side", set.repsPerSide)
    val pulls = PlanMath.repCount(set, plan.handMode)
    return L10n.tr("%d %s", pulls, L10n.tr(if (pulls == 1) "pull" else "pulls"))
}

// MARK: - Rhythm and cadence

@Composable
private fun RhythmCard(plan: SessionPlan, draft: RoutineDraft, setCount: Int) {
    val palette = LocalGripPalette.current
    Surface(
        shape = RoundedCornerShape(Metrics.radiusCard),
        color = palette.card,
        modifier = Modifier.fillMaxWidth(),
    ) {
        Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(14.dp)) {
            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                CapsLabel(tr("RHYTHM"))
                Text(
                    rhythmLine(plan, setCount),
                    style = MaterialTheme.typography.bodyMedium.copy(fontFeatureSettings = "tnum"),
                    fontWeight = FontWeight.Medium,
                    color = palette.inkPrimary,
                )
                Text(
                    plan.handMode.displayName,
                    style = MaterialTheme.typography.bodySmall,
                    color = palette.inkSecondary,
                )
            }
            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                CapsLabel(tr("EVERY DAY"))
                Text(
                    cadenceLine(draft),
                    style = MaterialTheme.typography.bodyMedium.copy(fontFeatureSettings = "tnum"),
                    fontWeight = FontWeight.Medium,
                    color = palette.inkPrimary,
                )
            }
        }
    }
}

/// The ROUTINE's rhythm, inherited unless a set overrides (overrides are already in the counts
/// and estimate).
private fun rhythmLine(plan: SessionPlan, setCount: Int): String {
    val parts = mutableListOf(
        L10n.tr("%s hold", PlanMath.durationText(plan.holdSeconds)),
        L10n.tr("%s rest", PlanMath.durationText(plan.restSeconds)),
    )
    // With one set, "between sets" is a constant dressed as information.
    if (setCount > 1) parts.add(L10n.tr("%s between sets", PlanMath.durationText(plan.setBreakSeconds)))
    return parts.joinToString(" · ")
}

/// A WHENEVER routine is never owed, so it says so rather than quoting a number.
private fun cadenceLine(draft: RoutineDraft): String {
    if (draft.isOnDemand) return L10n.tr("Whenever you're fresh")
    return when (val n = maxOf(1, draft.sessionsPerDay)) {
        1 -> L10n.tr("Once a day")
        2 -> L10n.tr("Twice a day")
        else -> L10n.tr("%d× a day", n)
    }
}

// MARK: - The two honesty notes

/// Each shown only when TRUE of this routine: noise footnotes teach people to stop reading them.
@Composable
private fun Notes(sets: List<SetPlan>, plan: SessionPlan) {
    val palette = LocalGripPalette.current
    /// `PlanMath.targetBand` precedence: a set with typed kilograms never reaches its percentage.
    val hasPercent = sets.any { it.targetBand == null && PlanMath.targetPercent(it, plan) != null }
    val hasKilograms = sets.any { it.targetBand != null }
    // Guarded around the STACK: an empty column is still a child, and the 18 dp spacing would
    // leave a gap under the (common) routine prescribing no load.
    if (!hasPercent && !hasKilograms) return

    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        if (hasPercent) {
            Text(
                tr("Percentage targets use your saved maxes. These may no longer reflect your current strength."),
                style = MaterialTheme.typography.bodySmall,
                color = palette.inkTertiary,
            )
        }
        if (hasKilograms) {
            Text(
                tr("Some fixed weight targets were set by the sender. Review them for your own training."),
                style = MaterialTheme.typography.bodySmall,
                color = palette.inkTertiary,
            )
        }
    }
}

File: Sources/UI/RoutineImportSheet.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

/// A routine decoded from a `getagrip://` link, waiting to be looked at. `Identifiable`
/// because `.sheet(item:)` is what presents it, with a fresh `id` per claim from the
/// store's inbox.
struct ImportRequest: Identifiable {
    let id = UUID()
    let draft: RoutineDraft
}

/// The other side of a QR code: somebody else's routine, read out in full, before it is
/// yours.
///
/// It is a PREVIEW, not an editor. Nothing here is adjustable, and that is the honest
/// shape — a routine you have not accepted yet is not a routine you can edit, and fields
/// would be asking a stranger's plan to be corrected before it has been read. Everything
/// in it is one tap from editable the moment it lands: the card it becomes opens the
/// builder on its own plan row.
///
/// **Percentage targets are deliberately NOT translated.** They resolve against the
/// READER's maxes, per hand, at the moment a session starts — which is the entire reason
/// this app prescribes fractions rather than kilograms, and it means the same code
/// prescribes the right load for two people with very different fingers. The two
/// footnotes exist for the cases where that is not the whole story.
struct RoutineImportSheet: View {
    @Environment(\.weightUnit) private var weightUnit
    /// NORMALIZED at init — see the initializer. Read as a value; the sheet observes no
    /// store except for the one save it performs.
    let draft: RoutineDraft

    @Environment(TemplateStore.self) private var templates
    @Environment(\.dismiss) private var dismiss
    @Environment(\.dynamicTypeSize) private var typeSize

    /// Whether THIS sheet's add failed — a local flag, not a read of
    /// `templates.saveError`: keyed to the store field, the sheet opened already
    /// wearing the accusation whenever an earlier, unrelated write had failed and its
    /// alert had not been read yet.
    @State private var saveFailed = false

    /// Folded ONCE, not per body evaluation: `RoutineSummary(previewing:)` walks the
    /// whole rep sequence three times, and it also mints an id, so a computed property
    /// would re-fold a fifty-set routine on every scroll frame and change identity while
    /// doing it.
    private let summary: RoutineSummary

    init(draft: RoutineDraft) {
        // Normalized HERE so the preview shows the routine that will actually land: an
        // empty name becomes the house default on the way in, an emptied set is dropped,
        // and inheritance is consolidated. Previewing the raw draft and saving the
        // normalized one is exactly how a preview and the card it becomes disagree.
        // `normalized` is idempotent, so the store's own pass costs nothing.
        let clean = draft.normalized
        self.draft = clean
        self.summary = RoutineSummary(previewing: clean)
    }

    private var plan: SessionPlan { draft.plan }
    /// `executable`, like every other fold in the app: a set with no pulls in it is a row
    /// the author emptied out, not a rest.
    private var sets: [SetPlan] { draft.plan.executable.sets }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 18) {
                    header
                    planCard
                    rhythmCard
                    notes
                }
                .padding(.horizontal, Metrics.hPadding)
                .padding(.top, 12)
                .padding(.bottom, 28)
                .frame(maxWidth: Metrics.maxContentWidth)
                .frame(maxWidth: .infinity)
            }
            // ALWAYS via `.background {}`, never as a ZStack sibling.
            .background { AppBackground() }
            .scrollBounceBehavior(.basedOnSize)
            .scrollEdgeEffectStyle(.soft, for: .bottom)
            .navigationTitle("Shared routine")
            .navigationBarTitleDisplayMode(.inline)
            // The decision lives in the safe area rather than at the foot of the
            // document: the payload allows fifty sets, and a primary action that has to
            // be scrolled to is a primary action people do not find.
            .safeAreaInset(edge: .bottom) { actions }
        }
    }

    // MARK: - Identity

    /// The name this routine will actually LAND under. The store deconflicts on save
    /// (`uniqueName`), and two people keeping the shipped default name is the common
    /// case for a shared routine — a preview promising "Daily no-hangs" four seconds
    /// before the card says "Daily no-hangs 2" is the preview and the card disagreeing.
    /// Computed per body pass, so a routine created behind the sheet is still counted.
    private var landingName: String {
        templates.plannedImportName(for: summary.name)
    }

    private var header: some View {
        HStack(alignment: .top, spacing: 12) {
            // The same derived mark Today's card wears, so the routine is recognisable
            // before and after it is yours.
            EdgeMark(fingers: summary.signatureFingers ?? .four,
                     rungTint: PlanMath.IntensityBand.band(for: summary.peakIntensity).tint)
                .padding(.top, 4)

            VStack(alignment: .leading, spacing: 4) {
                Text(landingName)
                    .font(.system(.title2, weight: .semibold))
                    .foregroundStyle(Ink.primary)
                    .fixedSize(horizontal: false, vertical: true)
                Text(summary.metaLine)
                    .font(.system(.footnote))
                    .monospacedDigit()
                    .foregroundStyle(Ink.secondary)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(landingName)
        // The rung's colour is invisible to VoiceOver and to greyscale, so the number
        // it stands for is spoken — the same pairing `RoutineCard` makes.
        .accessibilityValue(summary.metaLine + intensitySuffix)
    }

    private var intensitySuffix: String {
        guard let peak = summary.peakIntensity else { return "" }
        return String(localized: ". Peak target \(Int((peak * 100).rounded())) percent of max.")
    }

    // MARK: - The plan

    private var planCard: some View {
        MaterialCard {
            VStack(alignment: .leading, spacing: 10) {
                CapsLabel(String(localized: "WHAT YOU'LL PULL"))
                ForEach(Array(sets.enumerated()), id: \.offset) { index, set in
                    setRow(set)
                    if index < sets.count - 1 {
                        Divider().overlay(Ink.tertiary.opacity(0.14))
                    }
                }
            }
        }
    }

    /// Two layouts, forked at `.accessibility1` — the same fork `RoutineCard`'s plan
    /// row makes, and for the same measured reason: side by side, the scaled glyph plus
    /// a priority-protected count left the grip sentence ~22 pt of a 330 pt row at the
    /// app's type ceiling, on the one screen whose job is stating the grip. Big text is
    /// served by words stacked in full width; the glyph is decoration it can spare.
    @ViewBuilder
    private func setRow(_ set: SetPlan) -> some View {
        if typeSize >= .accessibility1 {
            VStack(alignment: .leading, spacing: 3) {
                Text(set.grip.line)
                    .font(.system(.subheadline, weight: .semibold))
                    .foregroundStyle(Ink.primary)
                    .fixedSize(horizontal: false, vertical: true)
                Text(repText(set))
                    .font(.system(.footnote, weight: .medium))
                    .monospacedDigit()
                    .foregroundStyle(Ink.secondary)
                    .fixedSize(horizontal: false, vertical: true)
                setDetail(set)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(setAccessibilityLabel(set))
        } else {
            HStack(alignment: .top, spacing: 12) {
                FingerGlyph(fingers: set.grip.fingers, position: set.grip.position,
                            dot: 7, gap: 3)
                    .padding(.top, 3)

                VStack(alignment: .leading, spacing: 2) {
                    Text(set.grip.line)
                        .font(.system(.subheadline, weight: .semibold))
                        .foregroundStyle(Ink.primary)
                        // No line limit: this sentence is what the screen exists to
                        // state, so it wraps rather than truncates.
                        .fixedSize(horizontal: false, vertical: true)
                    setDetail(set)
                }

                Spacer(minLength: 8)

                Text(repText(set))
                    .font(.system(.footnote, weight: .medium))
                    .monospacedDigit()
                    .foregroundStyle(Ink.secondary)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(setAccessibilityLabel(set))
        }
    }

    /// What this set does DIFFERENTLY — its own timing, its own load. A stranger's
    /// routine owes you these before you accept it: a typed 40 kg band you meet for the
    /// first time in the runner is exactly the surprise this sheet exists to prevent,
    /// and a per-set 3 s hold explains why the estimate above disagrees with the rhythm
    /// line below. Nothing renders for the common set that inherits everything.
    @ViewBuilder
    private func setDetail(_ set: SetPlan) -> some View {
        if let line = setDetailLine(set) {
            Text(line)
                .font(.system(.footnote))
                .monospacedDigit()
                .foregroundStyle(Ink.secondary)
                .fixedSize(horizontal: false, vertical: true)
        }
    }

    private func setDetailLine(_ set: SetPlan) -> String? {
        var parts: [String] = []
        if let hold = set.holdSeconds { parts.append(String(localized: "\(PlanMath.durationText(hold)) hold")) }
        if let rest = set.restSeconds { parts.append(String(localized: "\(PlanMath.durationText(rest)) rest")) }
        if let band = set.targetBand {
            parts.append(String(localized: "\(weightText(band.lowerBound))–\(weightText(band.upperBound)) \(weightUnit.symbol) target"))
        } else if let band = set.targetPercentBand {
            let lo = Int((band.lowerBound * 100).rounded())
            let hi = Int((band.upperBound * 100).rounded())
            parts.append(lo == hi ? String(localized: "\(hi) % of max") : String(localized: "\(lo)–\(hi) % of max"))
        }
        return parts.isEmpty ? nil : parts.joined(separator: " · ")
    }

    private func weightText(_ kg: Double) -> String {
        weightUnit.number(kg)
    }

    private func setAccessibilityLabel(_ set: SetPlan) -> String {
        var label = String(localized: "\(set.grip.spoken). \(repText(set)).")
        if let detail = setDetailLine(set) { label += String(localized: " \(detail).") }
        return label
    }

    /// "6 per side" when the hands take turns, "6 pulls" when they are on the edge
    /// together. The pull count comes from `PlanMath.repCount` rather than a
    /// `× sideCount` written here — that multiplication has exactly one home in the app,
    /// and this is the screen where a silent factor of two would be believed.
    private func repText(_ set: SetPlan) -> String {
        guard plan.handMode.sideCount == 1 else { return String(localized: "\(set.repsPerSide) per side") }
        let pulls = PlanMath.repCount(set, mode: plan.handMode)
        return String(localized: "\(pulls) \(pulls == 1 ? String(localized: "pull") : String(localized: "pulls"))")
    }

    // MARK: - Rhythm and cadence

    private var rhythmCard: some View {
        MaterialCard {
            VStack(alignment: .leading, spacing: 14) {
                VStack(alignment: .leading, spacing: 6) {
                    CapsLabel(String(localized: "RHYTHM"))
                    Text(rhythmLine)
                        .font(.system(.subheadline, weight: .medium))
                        .monospacedDigit()
                        .foregroundStyle(Ink.primary)
                        .fixedSize(horizontal: false, vertical: true)
                    Text(plan.handMode.name)
                        .font(.system(.footnote))
                        .foregroundStyle(Ink.secondary)
                        .fixedSize(horizontal: false, vertical: true)
                }
                VStack(alignment: .leading, spacing: 6) {
                    CapsLabel(String(localized: "EVERY DAY"))
                    Text(cadenceLine)
                        .font(.system(.subheadline, weight: .medium))
                        .monospacedDigit()
                        .foregroundStyle(Ink.primary)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
        }
    }

    /// The ROUTINE's rhythm — the values every set inherits unless it overrides. A set
    /// that overrides is already folded into the pull counts and the estimate above.
    private var rhythmLine: String {
        var parts = [String(localized: "\(PlanMath.durationText(plan.holdSeconds)) hold"),
                     String(localized: "\(PlanMath.durationText(plan.restSeconds)) rest")]
        // A break "between sets" is a constant dressed as information when there is only
        // one set — the Live Activity drops "Set 1 of 1" for the same reason.
        if summary.setCount > 1 {
            parts.append(String(localized: "\(PlanMath.durationText(plan.setBreakSeconds)) between sets"))
        }
        return parts.joined(separator: " · ")
    }

    /// A WHENEVER routine has no daily target and is never owed, so it says so instead
    /// of quoting a number it does not mean.
    private var cadenceLine: String {
        if draft.isOnDemand { return String(localized: "Whenever you're fresh") }
        switch max(1, draft.sessionsPerDay) {
        case 1: return String(localized: "Once a day")
        case 2: return String(localized: "Twice a day")
        case let n: return String(localized: "\(n)× a day")
        }
    }

    // MARK: - The two honesty notes

    /// Both are shown only when they are TRUE of this routine — a footnote about
    /// kilograms under a routine that carries none is noise, and noise is what teaches
    /// people to stop reading footnotes.
    @ViewBuilder private var notes: some View {
        // Guarded around the STACK, not just inside it: an empty VStack is still a view,
        // and the document's 18 pt spacing would leave a block of nothing under a routine
        // that prescribes no load at all — which is most of them.
        if hasPercentTargets || hasKilogramTargets {
            VStack(alignment: .leading, spacing: 8) {
                if hasPercentTargets {
                    note(String(localized: "Percentage targets use your saved maxes. These may no longer reflect your current strength."))
                }
                if hasKilogramTargets {
                    note(String(localized: "Some fixed weight targets were set by the sender. Review them for your own training."))
                }
            }
        }
    }

    private func note(_ text: String) -> some View {
        Text(text)
            .font(.system(.footnote))
            .foregroundStyle(Ink.tertiary)
            .fixedSize(horizontal: false, vertical: true)
            .frame(maxWidth: .infinity, alignment: .leading)
    }

    /// `PlanMath.targetBand`'s precedence, read as a predicate: a set carrying typed
    /// kilograms never reaches its percentage, so it is not a percentage set.
    private var hasPercentTargets: Bool {
        sets.contains { $0.targetBand == nil && PlanMath.targetPercent($0, in: plan) != nil }
    }

    private var hasKilogramTargets: Bool {
        sets.contains { $0.targetBand != nil }
    }

    // MARK: - The decision

    private var actions: some View {
        VStack(spacing: 4) {
            if saveFailed {
                // The sheet STAYS OPEN on a rollback: dismissing on failure loses the
                // code as well as the routine, and rescanning is somebody else's phone
                // away. Same inline treatment the builder gives its own failed save.
                Text("That routine couldn't be saved just now — nothing was added. Try again.")
                    .font(.system(.footnote, weight: .medium))
                    .foregroundStyle(Accent.alarm)
                    .fixedSize(horizontal: false, vertical: true)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.bottom, 6)
            }

            PrimaryGlassButton(title: String(localized: "Add to my routines"),
                               systemImage: "plus",
                               tint: Accent.graphite) {
                add()
            }

            // Quiet, and never destructive-looking: declining a routine costs nothing
            // and undoes nothing.
            Button("Not now") { dismiss() }
                .buttonStyle(PressFeedbackButtonStyle())
                .font(.system(.footnote, weight: .semibold))
                .foregroundStyle(Accent.graphite)
                // Drawn as one line of footnote, hit as 44 — and `contentShape` is
                // mandatory rather than tidy, since padding contributes nothing to
                // SwiftUI's default hit area.
                .frame(maxWidth: .infinity, minHeight: 44)
                .contentShape(.rect)
        }
        .padding(.horizontal, Metrics.hPadding)
        .padding(.bottom, 4)
        .frame(maxWidth: Metrics.maxContentWidth)
        .frame(maxWidth: .infinity)
    }

    /// The new card appears on Today by itself — the routine list there is a `@Query`,
    /// so nothing has to be handed back through the presentation.
    private func add() {
        guard templates.importRoutine(draft) != nil else {
            // Surfaced INLINE, and the store's copy of the failure is consumed — the
            // global "Couldn't save" alert watches the same field from the very view
            // presenting this sheet, and one rollback stated twice reads as two.
            saveFailed = true
            templates.saveError = nil
            return
        }
        saveFailed = false
        dismiss()
    }
}
